feat(curriculum): add Artifact Provenance Auditor workshop (#68860)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/blocks/workshop-artifact-provenance-auditor/6a57c331ae94fa4631b60871.md

---
id: 6a57c331ae94fa4631b60871
title: Step 18
challengeType: 1
dashedName: step-18
---

# --description--

An artifact's current gallery is its most recent location, which is the last item in its `locations` array.

Add a variable named `currentLocation` and assign it the last element of the artifact's `locations` array. Use the array's `length` to reach the last index.

# --hints--

You should declare a variable named `currentLocation` assigned to the last element of the artifact's `locations` array.

```js
const explorer = await __helpers.Explorer(code);

> assert.exists(explorer.allFunctions.buildSummary.variables.currentLocation);

Reviewer: ```suggestion
assert.exists(explorer.allFunctions.buildSummary?.variables.currentLocation);
```

console.log(getArtifactTitle(102));

function addTag(id, tag) {
  const artifact = collection[id];
  if (artifact && !artifact.tags.includes(tag)) {
    artifact.tags.push(tag);
  }
}

console.log("UPDATED TAGS");
addTag(101, "royal");
console.log(collection[101].tags);

function moveArtifact(id, gallery, year) {
  const artifact = collection[id];
  if (artifact) {
    artifact.locations.push({ gallery, year });
  }
}

console.log("UPDATED LOCATIONS");
moveArtifact(102, "Hall B", 2026);
console.log(collection[102].locations);

function toggleDisplayStatus(id) {
  const artifact = collection[id];
  if (artifact) {
    artifact.onDisplay = !artifact.onDisplay;
  }
}

console.log("DISPLAY STATUS");
console.log(collection[102].onDisplay);
toggleDisplayStatus(102);
console.log(collection[102].onDisplay);

function updateCurator(id, name) {
  const artifact = collection[id];
  if (artifact) {
    artifact.curator.name = name;
  }
}

console.log("UPDATED CURATOR");
updateCurator(101, "Fran Sinclair");
console.log(collection[101].curator.name);

function buildSummary(id) {
  const artifact = collection[id];
  if (!artifact) {
    return "Artifact not found";
  }
--fcc-editable-region--

> 

Reviewer: Indent the empty line b/w the markers with 2 spaces.

---
id: 6a57c331ae94fa4631b60871
title: Step 18
challengeType: 1
dashedName: step-18
---

# --description--

> An artifact's current gallery is its most recent location, which is the last item in its `locations` array.

Reviewer: ```suggestion
An artifact's current gallery is its most recent location. You can find this by checking the last item in its `locations` array.
```

> Add a variable named `currentLocation` and assign it the last element of the artifact's `locations` array. Use the array's `length` to reach the last index.

Reviewer: ```suggestion
Declare a `currentLocation` variable and assign it the last element of the artifact's `locations` array. You can use the array's `length` property to access the last index.
```

# --hints--

You should declare a variable named `currentLocation` assigned to the last element of the artifact's `locations` array.

```js

> assert.match(__helpers.removeJSComments(code), /(?:const|let|var)\s+currentLocation\s*=\s*artifact\s*\.\s*locations\s*\[\s*artifact\s*\.\s*locations\s*\.\s*length\s*-\s*1\s*\]/);

Reviewer: You can replace this regex with AST helpers. Please refer to the [AST Helpers documentation.](https://contribute.freecodecamp.org/curriculum-help/#typescript-helpers).
